tla-types/src/main/scala/at/forsyte/apalache/tla/typecomp/ScopedBuilder.scala

```diff
@@ -11,25 +11,118 @@ import scalaz.Scalaz._
+ * [[BuilderUtil.signatureMapEntry signatureMapEntry]] provides a utility method, for constructing such entries from
+ * partial functions (the implementation calls [[BuilderUtil.checkForArityException checkForArityException]] and
+ * [[BuilderUtil.completePartial completePartial]] in the background, to produce sensible error messages, if
+ * `ScopedBuilder.foo` is used with type-incorrect arguments). In our case:
+ * {{{
+ *   val fooSig = signatureMapEntry(foo, { case Seq(a, FunT1(aa, b)) if a == aa => b })
+ * }}}
+ * We then need to make sure that [[TypeComputationFactory]] adds the [[SignatureMap]] from `FooOperSignatures` to its
+ * `knownSignatures`.
+ *
+ * ==Implementing the type-safe builder method==
+ *
+ * The next step is adding (or extending) a builder in [[unsafe]], `UnsafeFooOperBuilder`. This builder should declare a
+ * method `foo(arg1: T1, ..., argN: TN): TlaEx`. The number of arguments depends on the particular operator, and is
+ * usually equal to the TLA+ arity of the operator, though this is not always the case. Typically, the argument types
+ * are `TlaEx` as well. An exception to this are, for example, operators which take type-hints. Take
+ * [[unsafe.UnsafeApalacheInternalBuilder.notSupportedByModelChecker notSupportedByModelChecker]]; it requires a
+ * `String` and a [[TlaType1]] argument, even though the TLA+ operator `NotSupportedByModelChecker` requires exactly one
+ * argument (of type `Str`).
+ *
+ * In our case, `foo` will need to accept 2 generic [[TlaEx]] arguments, so it will look like
+ * {{{
+ *   def foo(x: TlaEx, y: TlaEx): TlaEx = ...
+ * }}}
+ * The implementation depends on whether the operator has an associated signature. If yes, we just need to use
+ * [[unsafe.ProtoBuilder.buildBySignatureLookup buildBySignatureLookup]]; All it needs are the operator (`foo`) and a
+ * number of [[TlaEx]] arguments. In our case:
+ * {{{
+ *   def foo(x: TlaEx, y: TlaEx): TlaEx = buildBySignatureLookup(foo, x, y)
+ * }}}
+ * If we did not have an associated signature, we would have to implement a custom [[TypeComputation]]. See
+ * [[unsafe.UnsafeVariantBuilder.variant variant]] or [[unsafe.UnsafeFunBuilder.rec rec]] for examples of this approach.
+ *
+ * ==Implementing the scope-safe builder method==
+ *
+ * Finally, we need to add the type-safe, scope-safe method to `FooOperBuilder` in [[subbuilder]]. This builder should
+ * declare a method `foo(arg1: TT1, ..., argN: TTN): TBuilderInstruction` and a local instance of `UnsafeFooOperBuilder`
+ * (typically named `unsafeBuilder`). The number of arguments should be the same as for the `UnsafeFooOperBuilder.foo`
+ * method (we also typically use the same argument names). Each argument that is typed as [[TlaEx]] in
+ * `UnsafeFooOperBuilder.foo` should be typed as [[TBuilderInstruction]] in `FooOperBuilder.foo` (though e.g. `String`
+ * and [[TlaType1]] arguments should have the same types in both). Typically, the argument types are `TlaEx` as well. In
```

Comment:
I don't understand what this sentence means here: "Typically, the argument types are `TlaEx` as well." Is this the same arguments that we say should be `TBuilderInstruction`? Does it mean the types of the arguments should be subtypes of both?

Reply:
I think I just accidentally copied this sentence from the unsafe builder section.